core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala

```diff
@@ -519,6 +521,20 @@ private[spark] class MesosClusterScheduler(
     val currentTime = new Date()
 
     stateLock.synchronized {
+      var it = offers.iterator()
+      while (it.hasNext) {
+        val offer = it.next()
+        val offerAttributes = toAttributeMap(offer.getAttributesList)
+        val meetsConstraints = matchesAttributeRequirements(driverOfferConstraints, offerAttributes)
+
+        if (!meetsConstraints) {
+          driver.declineOffer(offer.getId)
+          it.remove()
```

Comment:
I'd rather not mutate the underlying collection. Anyhow, this has no effect because the code below is using `currentOffers`, which is a copy of the original resources made earlier. I suggest you insert this filter at that point.